Avoid too long blocks of the datasource caused by DB failover

### Description

When a DB server suddenly goes down and fails over without sending RST packet, the JDBC connections pooled in the datasource remain kept. 
These connections will fail with a timeout error when the application attempts to use them next time.

To prevent this, Quarkus datasource support connection validation either in the background or at the time the connection is pulled from the pool. 
However, in the current implementation, no query timeout is set for the validation query, causing it to wait until the standard socket timeout is reached.
```
Apr 01 11:02:09 rhbk kc.sh[1786]: 2025-04-01 11:02:09,862 WARN  [io.quarkus.agroal.runtime.DataSources] (org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker) Connection validation failed: org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:399)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:517)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:434)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:356)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:341)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:317)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:312)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at io.agroal.pool.wrapper.StatementWrapper.execute(StatementWrapper.java:220)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at io.quarkus.agroal.runtime.DataSources$3.isValid(DataSources.java:404)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at io.agroal.pool.ConnectionHandler.isValid(ConnectionHandler.java:244)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at io.agroal.pool.ConnectionPool.performValidation(ConnectionPool.java:395)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at io.agroal.pool.ConnectionPool.borrowValidation(ConnectionPool.java:387)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at io.agroal.pool.ConnectionPool.getConnection(ConnectionPool.java:290)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at io.agroal.pool.DataSource.getConnection(DataSource.java:86)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at io.quarkus.hibernate.orm.runtime.customized.QuarkusConnectionProvider.getConnection(QuarkusConnectionProvider.java:23)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.internal.NonContextualJdbcConnectionAccess.obtainConnection(NonContextualJdbcConnectionAccess.java:46)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:126)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.getPhysicalConnection(LogicalConnectionManagedImpl.java:156)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.connection(StatementPreparerImpl.java:56)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$4.doPrepare(StatementPreparerImpl.java:151)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:180)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.prepareQueryStatement(StatementPreparerImpl.java:153)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.sql.exec.internal.StandardStatementCreator.createStatement(StandardStatementCreator.java:49)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.executeQuery(DeferredResultSetAccess.java:235)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.getResultSet(DeferredResultSetAccess.java:171)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.sql.results.jdbc.internal.JdbcValuesResultSetImpl.<init>(JdbcValuesResultSetImpl.java:74)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.resolveJdbcValuesSource(JdbcSelectExecutorStandardImpl.java:355)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.doExecuteQuery(JdbcSelectExecutorStandardImpl.java:137)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.executeQuery(JdbcSelectExecutorStandardImpl.java:102)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.sql.exec.spi.JdbcSelectExecutor.scroll(JdbcSelectExecutor.java:195)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.lambda$new$2(ConcreteSqmSelectQueryPlan.java:187)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.withCacheableSqmInterpretation(ConcreteSqmSelectQueryPlan.java:442)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.performScroll(ConcreteSqmSelectQueryPlan.java:370)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.query.sqm.internal.QuerySqmImpl.doScroll(QuerySqmImpl.java:456)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.query.spi.AbstractSelectionQuery.scroll(AbstractSelectionQuery.java:235)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.query.spi.AbstractSelectionQuery.stream(AbstractSelectionQuery.java:252)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.hibernate.query.spi.AbstractSelectionQuery.getResultStream(AbstractSelectionQuery.java:246)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.jpa.session.JpaUserSessionPersisterProvider.loadUserSession(JpaUserSessionPersisterProvider.java:334)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.sessions.infinispan.changes.JpaChangesPerformer.processUserSessionUpdate(JpaChangesPerformer.java:519)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.sessions.infinispan.changes.JpaChangesPerformer.lambda$registerChange$0(JpaChangesPerformer.java:72)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.sessions.infinispan.changes.PersistentUpdate.perform(PersistentUpdate.java:40)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.lambda$process$0(PersistentSessionsWorker.java:90)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.lambda$process$1(PersistentSessionsWorker.java:90)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:274)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:393)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:273)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:263)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.lambda$process$3(PersistentSessionsWorker.java:89)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.common.util.Retry.executeWithBackoff(Retry.java:102)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.common.util.Retry.executeWithBackoff(Retry.java:93)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.process(PersistentSessionsWorker.java:86)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.run(PersistentSessionsWorker.java:69)
Apr 01 11:02:09 rhbk kc.sh[1786]: Caused by: java.net.SocketTimeoutException: Read timed out
Apr 01 11:02:09 rhbk kc.sh[1786]:         at java.base/sun.nio.ch.NioSocketImpl.timedRead(NioSocketImpl.java:288)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at java.base/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:314)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at java.base/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:355)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at java.base/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:808)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at java.base/java.net.Socket$SocketInputStream.read(Socket.java:966)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:192)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:159)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:144)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:76)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.core.PGStream.receiveChar(PGStream.java:476)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2174)
Apr 01 11:02:09 rhbk kc.sh[1786]:         at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:372)
Apr 01 11:02:09 rhbk kc.sh[1786]:         ... 52 more
```

When the DB server fails over, all pooled connections in the datasource become broken. 
As the validation is performed sequentially, if there are many pooled connections, leading the application to be blocked for the duration of (socket timeout) * (number of connections).
As a result, the Quarkus application remains unavailable for a long time, even after the DB server failover has completed.

### Implementation ideas

### 1. Add the ability to configure a query timeout for validation queries in Quarkus

Agroal, the data source implementation used by Quarkus, has a validator that supports issuing validation queries with query timeouts,
https://github.com/agroal/agroal/blob/d53895968cdca22fac01dc674dc584845dd0261e/agroal-api/src/main/java/io/agroal/api/configuration/AgroalConnectionPoolConfiguration.java#L206-L217

but it is currently not being used in Quarkus.
https://github.com/quarkusio/quarkus/blob/3.21.0/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSources.java#L314
https://github.com/quarkusio/quarkus/blob/3.21.0/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSources.java#L324-L336

We should have a property "quarkus.datasource.jdbc.validation-query-timeout" or something like that.

Note: JBoss EAP and ironjacamar supports this feature: https://issues.redhat.com/browse/JBEAP-18712

### 2. Add "use-fast-fail" option

When Quarkus fails to validate a connection, it attempts to obtain the next connection in the pool, which can result in a prolonged unavailable state if all connections in the pool are corrupted by a DB failover.
On the other hand, JBoss EAP provided an option ("use-fast-fail") to create a new connection instead of obtaining the next connection from the pool when connection validation fails.
The same functionality is needed for Quarkus.

--- a/extensions/agroal/deployment/src/test/java/io/quarkus/agroal/test/QueryLoggingH2DBEventListener.java
+++ b/extensions/agroal/deployment/src/test/java/io/quarkus/agroal/test/QueryLoggingH2DBEventListener.java
@@ -1,0 +1,50 @@
+package io.quarkus.agroal.test;
+
+import java.io.FileNotFoundException;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.logging.Handler;
+import java.util.logging.SimpleFormatter;
+import java.util.stream.Collectors;
+
+import org.apache.http.NameValuePair;
+import org.apache.http.client.utils.URLEncodedUtils;
+import org.h2.api.DatabaseEventListener;
+import org.jboss.logmanager.Logger;
+import org.jboss.logmanager.handlers.FileHandler;
+
+public class QueryLoggingH2DBEventListener implements DatabaseEventListener {
+
+    Logger logger = Logger.getLogger(QueryLoggingH2DBEventListener.class.getName());
+
+    @Override
+    public void init(String url) {
+        int index = url.indexOf("?");
+        if (index == -1) {
+            throw new IllegalArgumentException("jdbc uri doesn't contain queryLog param: " + url);
+        }
+        String query = url.substring(index + 1);
+        Map<String, String> params = URLEncodedUtils.parse(query, StandardCharsets.UTF_8).stream()
+                .collect(Collectors.toMap(NameValuePair::getName, NameValuePair::getValue));
+        String filePath = params.get("queryLog");
+        if (filePath == null) {
+            throw new IllegalArgumentException("jdbc uri doesn't contain queryLog param: " + url);
+        }
+
+        try {
+            Handler handler = new FileHandler(filePath, false);
+            handler.setFormatter(new SimpleFormatter());
+            logger.addHandler(handler);
+        } catch (FileNotFoundException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public void setProgress(int state, String name, long x, long max) {
+        if (state == DatabaseEventListener.STATE_STATEMENT_START) {
+            logger.info(name);
+        }
+    }
+}

--- a/extensions/agroal/deployment/src/test/java/io/quarkus/agroal/test/ValidationQueryTest.java
+++ b/extensions/agroal/deployment/src/test/java/io/quarkus/agroal/test/ValidationQueryTest.java
@@ -1,0 +1,42 @@
+package io.quarkus.agroal.test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.File;
+import java.sql.Connection;
+import java.sql.SQLException;
+
+import jakarta.inject.Inject;
+
+import org.junit.jupiter.api.*;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.agroal.api.AgroalDataSource;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ValidationQueryTest {
+
+    private static final String testLogPath = "target/validation-query-test.log";
+
+    //tag::injection[]
+    @Inject
+    AgroalDataSource defaultDataSource;
+    //end::injection[]
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withConfigurationResource("application-datasource-with-validation.properties")
+            .overrideRuntimeConfigKey("quarkus.datasource.jdbc.url",
+                    "jdbc:h2:tcp://localhost/mem:default?queryLog=%s;DATABASE_EVENT_LISTENER=io.quarkus.agroal.test.QueryLoggingH2DBEventListener" //Register QueryLoggingH2DBEventListener
+                            .formatted(testLogPath));
+
+    @Test
+    public void testQueryTimeoutIsApplied() throws SQLException {
+
+        //Test connection is acquirable
+        try (Connection connection = defaultDataSource.getConnection()) {
+            //nop
+        }
+        assertThat(new File(testLogPath)).content().contains("SET QUERY_TIMEOUT ?");
+    }
+}

--- a/extensions/agroal/deployment/src/test/resources/application-datasource-with-validation.properties
+++ b/extensions/agroal/deployment/src/test/resources/application-datasource-with-validation.properties
@@ -1,0 +1,10 @@
+quarkus.datasource.db-kind=h2
+
+quarkus.datasource.jdbc.min-size=1
+quarkus.datasource.jdbc.initial-size=1
+quarkus.datasource.jdbc.max-size=1
+quarkus.datasource.jdbc.validate-on-borrow=true
+quarkus.datasource.jdbc.validation-query-sql=select 1
+quarkus.datasource.jdbc.validation-query-timeout=3s
+
+quarkus.datasource.devservices.enabled=false

--- a/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceJdbcRuntimeConfig.java
+++ b/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceJdbcRuntimeConfig.java
@@ -118,6 +118,11 @@ public interface DataSourceJdbcRuntimeConfig {
     Optional<String> validationQuerySql();
 
     /**
+     * The timeout for the connection validation query
+     */
+    Optional<Duration> validationQueryTimeout();
+
+    /**
      * Forces connection validation prior to acquisition (foreground validation) regardless of the idle status.
      * <p>
      * Because of the overhead of performing validation on every call, it’s recommended to rely on default idle validation

--- a/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSources.java
+++ b/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSources.java
@@ -311,7 +311,12 @@ public class DataSources {
         }
 
         // Connection management
-        poolConfiguration.connectionValidator(ConnectionValidator.defaultValidator());
+        if (dataSourceJdbcRuntimeConfig.validationQueryTimeout().isPresent()) {
+            poolConfiguration.connectionValidator(ConnectionValidator
+                    .defaultValidatorWithTimeout((int) dataSourceJdbcRuntimeConfig.validationQueryTimeout().get().toSeconds()));
+        } else {
+            poolConfiguration.connectionValidator(ConnectionValidator.defaultValidator());
+        }
         if (dataSourceJdbcRuntimeConfig.acquisitionTimeout().isPresent()) {
             poolConfiguration.acquisitionTimeout(dataSourceJdbcRuntimeConfig.acquisitionTimeout().get());
         }
@@ -326,6 +331,9 @@ public class DataSources {
                 @Override
                 public boolean isValid(Connection connection) {
                     try (Statement stmt = connection.createStatement()) {
+                        if (dataSourceJdbcRuntimeConfig.validationQueryTimeout().isPresent()) {
+                            stmt.setQueryTimeout((int) dataSourceJdbcRuntimeConfig.validationQueryTimeout().get().toSeconds());
+                        }
                         stmt.execute(validationQuery);
                         return true;
                     } catch (Exception e) {

